### PR TITLE
Daml lf type safty (Party & PackageId)

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
In this PR we split the Simple String in two different types: 
 * Party 
 * PackageId

This fixes part of  #705

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/docs/source/support/release-notes.rst), if appropriate

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/MatchingStringModule.scala [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+sealed abstract class MatchingStringModule {
+  type T <: String
+
+  def fromString(s: String): Either[String, T]
+
+  @throws[IllegalArgumentException]
+  def assertFromString(s: String): T =
+    fromString(s).fold(e => throw new IllegalArgumentException(e), identity)
+
+  def unapply(x: T): Option[String] = Some(x)
```
Comment (S11001001) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/MatchingStringModule.scala]: Totality checking actually understands the implication of this:

```suggestion
  def unapply(x: T): Some[String] = Some(x)
```

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Ref.scala [hunk 2]
```diff
@@ -157,17 +115,30 @@ object Ref {
    * specified package. */
   case class Identifier(packageId: PackageId, qualifiedName: QualifiedName)
 
+  /* Choice name in a template. */
+  type ChoiceName = String
+
+  type ModuleName = DottedName
+  val ModuleName = DottedName
+
+  /** PackageId/Party are non empty US-ASCII strings built with letters, digits, space, minus and,
```
Comment (S11001001) [inline daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Ref.scala]:
```suggestion
  /** Party are non empty US-ASCII strings built with letters, digits, space, minus and,
```

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 3]
```diff
@@ -213,7 +213,7 @@ object SBuiltin {
   final case object SBToQuotedTextParty extends SBuiltin(1) {
     def execute(args: util.ArrayList[SValue], machine: Machine): Unit = {
       val v = args.get(0).asInstanceOf[SParty]
-      machine.ctrl = CtrlValue(SText(s"'${v.value.underlyingString}'"))
+      machine.ctrl = CtrlValue(SText(s"'${v.value}'"))
```
Comment (S11001001) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala]: To make sure the implied `toString` is benign in the face of future changes:

```suggestion
      machine.ctrl = CtrlValue(SText(s"'${v.value: String}'"))
```

I've discussed with @leo-da doing this as a rule, and replacing `s` in the future with one that only accepts some types for interpolation; for now this is a good stopgap.

File: daml-lf/parser/src/main/scala/com/digitalasset/daml/lf/testing/parser/Parsers.scala [hunk 4]
```diff
@@ -23,13 +23,16 @@ private[parser] object Parsers extends scala.util.parsing.combinator.Parsers {
 
   val id: Parser[String] = accept("Identifier", { case Id(s) => s })
   val text: Parser[String] = accept("Text", { case Text(s) => s })
-  val simple: Parser[Ref.SimpleString] = accept("SimpleString", { case SimpleString(s) => s })
+  val pkgId: Parser[Ref.PackageId] = accept("PackageId", {
```
Comment (S11001001) [inline daml-lf/parser/src/main/scala/com/digitalasset/daml/lf/testing/parser/Parsers.scala]:
```suggestion
  val pkgId: Parser[Ref.PackageId] = accept("PackageId", Function unlift {
    case SimpleString(s) => Ref.PackageId.fromString(s).toOption
    case _ => None
```

File: language-support/scala/codegen/src/main/scala/com/digitalasset/codegen/lf/EventDecoderGen.scala [hunk 5]
```diff
@@ -43,12 +43,12 @@ object EventDecoderGen {
     // the ledger api still uses names with only dots in them, while QualifiedName.toString
     // separates the module and the name in the module with colon.
     def legacyDottedName(identifier: Identifier) = {
-      val Identifier(SimpleString(packageId), QualifiedName(module, name)) = identifier
-      s"${module.dottedName: String}.${name.dottedName: String}@${packageId: String}"
+      val Identifier(packageId, QualifiedName(module, name)) = identifier
+      s"${module.dottedName: String}.${name.dottedName: String}@$packageId"
```
Comment (S11001001) [inline language-support/scala/codegen/src/main/scala/com/digitalasset/codegen/lf/EventDecoderGen.scala]: it's not paranoia if scala is really out to get you

```suggestion
      s"${module.dottedName: String}.${name.dottedName: String}@${packageId: String}"
```

File: ledger/ledger-api-common/src/main/scala/com/digitalasset/platform/server/api/services/grpc/GrpcTransactionService.scala [hunk 6]
```diff
@@ -83,9 +83,9 @@ class GrpcTransactionService(
       visibleTx: VisibleTransaction,
       offset: String,
       verbose: Boolean): TransactionTree = {
-    val mappedDisclosure = visibleTx.disclosureByNodeId.map {
-      case (k, v) => k -> v.map(_.underlyingString)
-    }
+    val mappedDisclosure =
+      visibleTx.disclosureByNodeId.transform((_, v) => v.toSet[String])
```
Comment (S11001001) [inline ledger/ledger-api-common/src/main/scala/com/digitalasset/platform/server/api/services/grpc/GrpcTransactionService.scala]: 👍 👍

File: ledger/ledger-api-common/src/test/suite/scala/com/digitalasset/platform/ledger/daml/runtime/PackageParserTest.scala [hunk 7]
```diff
@@ -17,14 +17,14 @@ class PackageParserTest extends WordSpec with Matchers {
...
-        Ref.SimpleString
+      val Some(pkgId) =
+        Ref.PackageId
           .fromString("497b17f4f5148e0a7102a5a402c3e35e67aa67a663520089922c677819ee4875")
           .right
           .toOption
       val result = PackageParser.parseDarOrDalf(() => getDarStream)
       result should matchPattern {
-        case Right((`simpleString`, Ast.Package(_))) =>
+        case Right((pkgId, Ast.Package(_))) =>
```
Comment (S11001001) [inline ledger/ledger-api-common/src/test/suite/scala/com/digitalasset/platform/ledger/daml/runtime/PackageParserTest.scala]:
```suggestion
        case Right((`pkgId`, Ast.Package(_))) =>
```

File: ledger/participant-state-index/reference/src/main/scala/com/daml/ledger/participant/state/index/v1/impl/reference/TransactionFiltering.scala [hunk 8]
```diff
@@ -20,8 +19,6 @@ case class TransactionFiltering(filter: TransactionFilter) {
       .getOrElse(Map.empty)
       .map {
```
Comment (S11001001) [inline ledger/participant-state-index/reference/src/main/scala/com/daml/ledger/participant/state/index/v1/impl/reference/TransactionFiltering.scala]:
```suggestion
      // import scalaz.syntax.std.map._ top of file
      .mapKeys(k => Value.NodeId.unsafeFromIndex(k.toInt))
```

File: ledger/participant-state/reference/src/main/scala/com/daml/ledger/participant/state/v1/impl/reference/Ledger.scala [hunk 9]
```diff
@@ -318,7 +318,7 @@ class Ledger(timeModel: TimeModel, timeProvider: TimeProvider)(implicit mat: Act
   }
 
   def uploadArchive(archive: Archive): Unit = {
-    val (pkgId: PackageId, pkg: Ast.Package) = Decode.decodeArchive(archive)
+    val (pkgId, pkg) = Decode.decodeArchive(archive)
```
Comment (S11001001) [inline ledger/participant-state/reference/src/main/scala/com/daml/ledger/participant/state/v1/impl/reference/Ledger.scala]: 👍 👍 👍

File: navigator/backend/src/main/scala/com/digitalasset/navigator/console/commands/Package.scala [hunk 10]
```diff
@@ -21,7 +22,7 @@ case object Package extends SimpleCommand {
     for {
       arg1 <- args.headOption ~> "Missing <id> argument"
       ps <- state.getPartyState ~> s"Unknown party ${state.party}"
-      pack <- ps.packageRegistry.pack(arg1) ~> s"Unknown package $arg1"
+      pack <- ps.packageRegistry.pack(PackageId.assertFromString(arg1)) ~> s"Unknown package $arg1"
```
Comment (S11001001) [inline navigator/backend/src/main/scala/com/digitalasset/navigator/console/commands/Package.scala]:
```suggestion
      packId <- PackageId.fromString(arg1).toOption ~> s"Invalid <id> argument $arg1"
      pack <- ps.packageRegistry.pack(packId) ~> s"Unknown package $arg1"
```

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/MatchingStringModule.scala (hunk 1 above)
Comment (remyhaemmerle-da) [reply daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/MatchingStringModule.scala]: fixed

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala (hunk 3 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala]: done

File: language-support/scala/codegen/src/main/scala/com/digitalasset/codegen/lf/EventDecoderGen.scala (hunk 5 above)
Comment (remyhaemmerle-da) [reply language-support/scala/codegen/src/main/scala/com/digitalasset/codegen/lf/EventDecoderGen.scala]: done

File: ledger/ledger-api-common/src/test/suite/scala/com/digitalasset/platform/ledger/daml/runtime/PackageParserTest.scala (hunk 7 above)
Comment (remyhaemmerle-da) [reply ledger/ledger-api-common/src/test/suite/scala/com/digitalasset/platform/ledger/daml/runtime/PackageParserTest.scala]: done

File: navigator/backend/src/main/scala/com/digitalasset/navigator/console/commands/Package.scala (hunk 10 above)
Comment (remyhaemmerle-da) [reply navigator/backend/src/main/scala/com/digitalasset/navigator/console/commands/Package.scala]: done

File: daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Ref.scala (hunk 2 above)
Comment (remyhaemmerle-da) [reply daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Ref.scala]: done

File: daml-lf/parser/src/main/scala/com/digitalasset/daml/lf/testing/parser/Parsers.scala (hunk 4 above)
Comment (remyhaemmerle-da) [reply daml-lf/parser/src/main/scala/com/digitalasset/daml/lf/testing/parser/Parsers.scala]: done

**Pull request merged**